pennylane/optimize/adaptive.py

```diff
@@ -0,0 +1,209 @@
+        qml.apply(m)
+
+
+class AdaptiveOptimizer:
+    r"""Adaptive optimizer.
+
+    Quantum circuits can be built by adding gates
+    `adaptively <https://www.nature.com/articles/s41467-019-10988-2>`_. The adaptive optimizer
+    implements an algorithm that grows and optimizes an input quantum circuit by selecting and
+    adding gates from a user-defined collection of operators. The algorithm starts with adding all
+    the gates to the circuit and computing the circuit gradients with respect to the gate
+    parameters. The algorithm then retains the gate which has the largest gradient and optimizes its
+    parameter. The processes of growing the circuit is repeated until the computed gradients
+    converge within a given threshold. The adaptive optimizer can be used to implement algorithms
+    such as `ADAPT-VQE <https://www.nature.com/articles/s41467-019-10988-2>`_.
+
+    Args:
+        paramopt_steps (float): number of steps for optimizing the parameter of a selected gate
+
+    **Examples:**
+
+    This examples shows an implementation of the
+    `ADAPT-VQE <https://www.nature.com/articles/s41467-019-10988-2>`_ algorithm for building an
+    adaptive circuit for the :math:`\text{H}_3^+` cation.
+
+    >>> import pennylane as qml
+    >>> from pennylane import numpy as np
+
+    The molecule is defined and the Hamiltonian is computed with:
+
+    >>> symbols = ["H", "H", "H"]
+    >>> geometry = np.array([[0.01076341, 0.04449877, 0.0],
+    ...                      [0.98729513, 1.63059094, 0.0],
+    ...                      [1.87262415, -0.00815842, 0.0]], requires_grad=False)
+    >>> H, qubits = qml.qchem.molecular_hamiltonian(symbols, geometry, charge = 1)
+
+    The collection of gates to grow the circuit adaptively contntains aa singles and doubles
+    excitations:
+
+    >>> n_electrons = 2
+    >>> singles, doubles = qml.qchem.excitations(n_electrons, qubits)
+    >>> operator_pool = doubles + singles
+
+    An initial circuit preparing the Hartree-Fock state and returning the expectation value of the
+    Hamiltonian is defined:
+
+    >>> hf_state = qml.qchem.hf_state(n_electrons, qubits)
+    >>> dev = qml.device("default.qubit", wires=qubits)
+    >>> @qml.qnode(dev)
+    ... def circuit():
+    ...     qml.BasisState(hf_state, wires=range(qubits))
+    ...     return qml.expval(H)
+
+    The optimizer is instantiated and the circuit is created and optimized adaptively:
+
+    >>> opt = AdaptiveOptimizer()
+    >>> params = np.zeros(len(operator_pool))
+    >>> operator_pool = pool_gate(params, operator_pool)
+    >>> for i in range(len(operator_pool)):
+    ...     circuit, energy, gradient = opt.step_and_cost(circuit, operator_pool, drain_pool=True)
+    ...     print('Energy:', energy)
+    ...     print(qml.draw(circuit)())
+    ...     print('Gradient max:', gradient)
+    ...     print()
+    ...     if gradient < 1e-3:
+    ...         break
+
+    .. code-block :: pycon
+
+    Energy: -1.246549938420637
```

Comment:
@soranjh don't forget to look at the rendered docs! 

![image](https://user-images.githubusercontent.com/2959003/198047388-d2479ecf-242c-43df-8af0-ec9639fc355b.png)